### ISSUE TITLE
generated code triggers numerous shellcheck warnings

### ISSUE THREAD

**Issue by calid** [remkop/picocli#644]
This came up while working  on a PR for apache/groovy, see: https://github.com/apache/groovy/pull/896

Basically I was helping smash some [shellcheck](https://github.com/koalaman/shellcheck) warnings in the groovy bin scripts, including the completions scripts which I didn't realize were actually generated by picocli.

You can see the shellcheck warnings for the completion scripts [here](https://gist.github.com/calid/634258c25f9d4aa0c1327eff1c5b16e1#file-groovy_completion-shellcheck-out)

And the changes I made [here](https://github.com/calid/groovy/commit/b26ecbaac381e78646333db3719d680e5f8a8dd5)

The main issues I encountered are:

* syntax of the ArrContains function is dubious, and this function doesn't really seem necessary
* variables aren't properly localized and pollute the user's environment
* there are unused variables (minor)

**Comment by remkop** [comment]
Hi, thanks for raising this!

Localizing the variables is a good idea, should definitely do that.

The ArrContains function is used to support nested subcommands, like used in `grape`. It is used because it supports nested sub-subcommands to any level of depth; this is quite powerful. 
Can you clarify what you mean by "dubious syntax"? I would consider alternative mechanisms that still support nested sub-subcommands, can you propose something?

I'll see what we can do about the unused variables.

**Comment by calid** [comment]
A few dubious constructs here:

`for i in "${lArr1[@]}";{ [ -n "$i" ] && ((++tmp[$i]));}`

using curly braces intead of `do`/`done` is relying on an undocumented feature meant to support backward compatibility for... an even older undocumented feature:
https://stackoverflow.com/questions/38828762/linux-curly-brackets-instead-of-if-then-else-fi-do-done

Really should be using the standard syntax of 

`for ... ; do ... ; done`

---

`eval lArr1=("\"\${$1[@]}\"")`

I get nervous around eval.  It lends itself to abuse and has many, many pitfalls.  It should be reserved for the few cases it makes sense and is well-suited, e.g. reading in environment variables dumped by an external program or dynamically building an options string for a program, not for logic like declaring and initializing associative arrays. (also why the parens here? There's no reason to execute that in a subshell?)
https://mywiki.wooledge.org/BashFAQ/048

---

Finally, just the use of associative arrays in general instead of case statements is dubious in a completion script.  Pretty much the universal idiom in these scripts is to use `case` and switch on `$cur` and `$prev` (set from `COMP_WORDS`).

To your question about how to handle arbitrarily nested subcommands without an associative array, I would say since the code is being generated just generate the necessary case statements to the necessary level of depth.

I think git's own completion script is really the gold standard here, especially when it comes to handling complex nested subcommands.  I would refer to that as the 'right' way to do things in a completion script.
https://github.com/git/git/blob/master/contrib/completion/git-completion.bash

**Comment by remkop** [comment]
That is good feedback, thank you for the clarification!
Do you think you will be able to provide a pull request to fix (some of) these items?

**Comment by calid** [comment]
I can add it to my personal FOSS queue, but I'm pretty slammed with work and other projects so not sure when or if I would get to it, so if there's someone else who'd like to take this on I will not be offended.

As far as I know nothing is actually broken, so this is probably isn't all that urgent, although I do think there's great value to generating completion scripts that adhere to the common idioms.

**Comment by AlcaYezz** [comment]
I like to work on the completion script generator. Variable localization, unused variable and for/if syntax are straightforward to fix.

To get ride of the ArrContains function it need a different PR I think and I'm not confident enough to go to the git completion script. Instead I can propose something based on @calid change with a tree-like solution to keep subcommand working. You can found a POC [here](https://gist.github.com/AlcaYezz/d9fe9b4dbb881f68c81e6de02e9861f0)

**Comment by remkop** [comment]
@AlcaYezz Great, pull requests are very welcome!

If I understand correctly you are proposing to split up the work into two PRs: one to clean up the syntax (local vars, for/if etc) and a separate one to work on a replacement of the associative array. I like that idea a lot. 

I haven’t had a chance to try the completion script you linked to yet (away from PC now).

**Comment by AlcaYezz** [comment]
Yes that's it, to PR for this, the first one is on the way let me know if my proposal for the completion script fit your need so I can start it.

**Comment by remkop** [comment]
Linking to @AlcaYezz ‘s pull request: https://github.com/remkop/picocli/pull/671

**Comment by remkop** [comment]
Thank you, @calid and @AlcaYezz for raising this and for providing a PR.
I finally got around to improving this.

In addition to @AlcaYezz's improvements, I also rewrote the `ArrContains` function: it has been broken up into two functions that simply loop over an array. The use of `eval` and associative arrays has been eliminated. This should also make it possible for the autocompletion scripts to run on OSX (https://github.com/remkop/picocli/issues/396).
